Core/include/Acts/Utilities/KDTree.hpp

```diff
@@ -0,0 +1,679 @@
+    ///
+    /// @param r The range to search for.
+    /// @param f The mapping function to apply to matching elements.
+    virtual void rangeSearchMapDiscard(
+        const range_t &r,
+        std::function<void(const coordinate_t &, const Type &)> f) const = 0;
+
+    /// @brief Debugging method that gives a textual representation of a k-d
+    /// tree.
+    ///
+    /// It is probably best not to use this method in real code. It is designed
+    /// for debugging purposes only.
+    ///
+    /// @param d The depth of this node (for indentation).
+    virtual std::string toString(std::size_t d) const = 0;
+
+    /// @brief Determine the number of elements managed by this node.
+    ///
+    /// Conveniently, this number is always equal to the distance between the
+    /// begin iterator and the end iterator, so we can simply delegate to the
+    /// relevant standard library method.
+    ///
+    /// @return The number of elements below this node.
+    std::size_t size(void) const { return std::distance(m_begin_it, m_end_it); }
+
+    /// @brief The axis-aligned bounding box containing all elements in this
+    /// node.
+    ///
+    /// @return The minimal axis-aligned bounding box that contains all the
+    /// elements under this node.
+    const range_t &range(void) const { return m_range; }
+
+   protected:
+    /// @brief The start and end of the range of coordinate-value pairs under
+    /// this node.
+    const iterator_t m_begin_it, m_end_it;
+
+    /// @brief The axis-aligned bounding box of the coordinates under this
+    /// node.
+    const range_t m_range;
+  };
+
+  /// @brief A non-leaf node in the k-d tree.
+  ///
+  /// These nodes are not the direct parents of any coordinate-value pairs, but
+  /// rather they are the parents of two other nodes. This means that this type
+  /// of node represents a split of the candidate space across a particular
+  /// dimension.
+  class KDTreeNode final : public KDTreeAbstractNode {
+   public:
+    /// @brief Constract an internal node from a range of coordinate-value
+    /// pairs.
+    ///
+    /// The element range passed to this constructor is a subrange of the set
+    /// of coordinate-value pairs owned by the outer-most parent node. The
+    /// constructor rearranges the elements of this range.
+    ///
+    /// @param _b The iterator for the start of the range of elements.
+    /// @param _e The iterator for the end of the range of elements.
+    KDTreeNode(iterator_t _b, iterator_t _e) : KDTreeAbstractNode(_b, _e) {
+      // This constant determines the maximum number of elements where we still
+      // calculate the exact median of the values for the purposes of
+      // splitting. In general, the closer the pivot value is to the true
+      // median, the more balanced the tree will be. However, calculating the
+      // median exactly is an O(n log n) operation, while approximating it is
+      // an O(1) time.
+      constexpr std::size_t max_exact_median = 128;
```

Comment:
Where does the number 128 come from - did you time that this is a good cross-over point?

Reply:
So this is a very good question with a lot of backstory behind it. When you traverse a binary tree, the number of steps depends on how well balanced it is. In the best case, if your tree with _n_ nodes is perfectly balanced, it takes _O_(log _n_) steps, and in the worst case, if your tree is perfectly unbalanced (we call this a degenerate binary tree), then it takes _O_(_n_) steps. From this it follows that you would like your tree to be as balanced as possible.

However, creating a balanced tree is not always easy. In order to generate a perfectly balanced tree, you need to split each subrange exactly in the middle. This is why k-d trees are usually built using the median, because of course the median, by definition, splits the range in half. If you always pick the exact median, you will always end up with a well-balanced tree.

Now the problem is that selecting the median value is an _O_(_n_ log _n_) operation. This means that you're trading off time spent on constructing on time spent looking up values in the tree. And this is where the problem arises: the optimal cut-off point where.

Luckily, in the particular use case I am targeting, the values in most of the dimension are symmetrically distributed (I think); the phi angle is uniformly distributed, and the z axis is probably mostly symmetrical. For symmetrical distribution the mean value and the median value are of course pretty close. The radius is not symmetrical, and there could be a bit of an issue there. But the mean approach should work fairly well for this particular use case.

There is a third solution, which is the median-of-median strategy, where you approximate the median of a range of values by finding the median of medians of sub-ranges. However, I didn't think this would be noticeable enough to be worth implementing.

So, to summarize, there is a Pareto optimal front of where you should cut off your exact median computation, but it depends on a lot of things: how big is your data set; how many times are you accessing this data per tree constructed; how symmetric is the distribution of your data; how large are your search ranges; what is the computational cost of sorting on your platform; and quite a few more.

The value of 128 here was chosen arbitrarily here. I picked quite a low value, because as I mentioned the data I am working with is relatively symmetric, and therefore the performance loss of using an approximate median is a lot lower than it is for arbitrary data.

I was thinking of making this parameter a compile-time template or even a run-time configuration parameter, but I didn't think it would be worth it for now. I am happy to implement that though, if you'd like!

Reply:
Excellent answer, thank you - I learned something!